lib/main.dart

```diff
@@ -0,0 +1,364 @@
+            participant.info.isLocal ||
+            participant.id == _participantFocusPriority.firstOrNull) return;
+        _activeSpeakerToSet = participant.id;
+        if (_participantFocusPriority.length < 2) {
+          _updateActiveSpeaker();
+        } else {
+          _setActiveSpeakerTimer ??= Timer(const Duration(seconds: 2), _updateActiveSpeaker);
+        }
+      },
+      participantJoined: (participant) {
+        if (!_hasActiveCall || participant.info.isLocal || _participantFocusPriority.length >= 3) {
+          return;
+        }
+        setState(() => _participantFocusPriority.add(participant.id));
+        _updateSubscriptions();
+      },
+      participantLeft: (participant) => setState(() {
+        _participantFocusPriority.remove(participant.id);
+        final participants = widget.callClient.participants;
+        if (_participantFocusPriority.length >= min(3, participants.remote.length)) return;
+        for (final participantId in participants.remote.keys) {
+          if (!_participantFocusPriority.contains(participantId)) {
+            _participantFocusPriority.add(participantId);
+          }
+          if (_participantFocusPriority.length >= 3) break;
+        }
+        _updateSubscriptions();
+      }),
+      appMessageReceived: _handleAppMessage,
+      error: (error) => _scaffoldMessengerKey.currentState?.showSnackBar(SnackBar(content: Text(error))),
+    );
+  }
+
+  void _updateActiveSpeaker() {
+    final activeSpeakerToSet = _activeSpeakerToSet;
+    if (mounted && _hasActiveCall && activeSpeakerToSet != null) {
+      setState(() {
+        _participantFocusPriority
+          ..remove(activeSpeakerToSet)
+          ..insert(0, activeSpeakerToSet);
+        if (_participantFocusPriority.length > 3) {
+          _participantFocusPriority.removeRange(3, _participantFocusPriority.length);
+        }
+      });
+    }
+    _activeSpeakerToSet = null;
+    _setActiveSpeakerTimer = null;
+    _updateSubscriptions();
+  }
+
+  void _updateSubscriptions() {
+    widget.callClient.updateSubscriptions(
+      // Profile-wide settings updates are applied before changes to individual participant subscriptions.
+      forParticipantsWithProfiles: {
+        SubscriptionProfile.activeRemote: const SubscriptionSettingsUpdate.set(
+          profile: SubscriptionProfileUpdate.set(profile: SubscriptionProfile.base),
+        ),
+      },
+      forParticipants: {
+        for (final id in _participantFocusPriority)
+          id: const SubscriptionSettingsUpdate.set(
+            profile: SubscriptionProfileUpdate.set(
+              profile: SubscriptionProfile.activeRemote,
+            ),
+          ),
+      },
+    );
+  }
+
+  void _handleAppMessage(final String data, final ParticipantId from) {
+    _handleParsedAppMessage(AppMessage.fromJson(jsonDecode(data)), from);
+  }
+
+  void _handleParsedAppMessage(final AppMessage data, final ParticipantId from) {
+    if (!_hasActiveCall) return;
```

Comment:
For my own understanding: when would this be the case? Usually an app message can only be received when the user is connected to a call. Is there some async logic here that means the user may disconnect before the message they received is handled?

Reply:
While everything goes well, this should never happen, but this is a distributed system. I don't think we have any hard guarantees that all messages/actions will always be delivered/taken in order. As such, a delayed message could conceivably arrive after we start disconnecting. Ie., our local state has been set to `leaving` but that has not yet taken effect in the SFU.